File: src/main/java/br/com/leonardoferreira/jirareport/service/DueDateService.java
package br.com.leonardoferreira.jirareport.service;

import br.com.leonardoferreira.jirareport.domain.DueDateType;
import br.com.leonardoferreira.jirareport.domain.embedded.DueDateHistory;
import br.com.leonardoferreira.jirareport.domain.vo.changelog.JiraChangelogItem;
import java.time.LocalDateTime;
import java.util.List;

public interface DueDateService {

    List<DueDateHistory> extractDueDateHistory(String dueDateCF, List<JiraChangelogItem> changelogItems);

    Long calcDeviationOfEstimate(List<DueDateHistory> dueDateHistories, LocalDateTime endDate, DueDateType dueDateType);

}

File: business-library/src/main/kotlin/br/com/jiratorio/service/DueDateService.kt
package br.com.jiratorio.service

import br.com.jiratorio.domain.changelog.FieldChangelog
import br.com.jiratorio.domain.entity.embedded.DueDateHistory
import br.com.jiratorio.extension.fromJiraToLocalDateTime
import br.com.jiratorio.extension.toLocalDate
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service
import java.time.LocalDateTime

@Service
class DueDateService {

    private val log = LoggerFactory.getLogger(javaClass)

    fun parseHistory(dueDateCF: String, fieldChangelog: Set<FieldChangelog>): List<DueDateHistory> {
        log.info("dueDateCF={}, fieldChangelog={}", dueDateCF, fieldChangelog)

        return fieldChangelog
            .filter { dueDateCF == it.field }
            .mapNotNull { parseDueDate(it.to, it.created) }
            .sortedBy { it.created }
    }

    private fun parseDueDate(dueDate: String?, created: LocalDateTime): DueDateHistory? =
        if (dueDate.isNullOrEmpty())
            null
        else
            DueDateHistory(
                created = created,
                dueDate = if (dueDate.length > 19)
                    dueDate.fromJiraToLocalDateTime().toLocalDate()
                else
                    dueDate.toLocalDate("[yyyy-MM-dd][dd/MM/yyyy]")
            )

}
